pkg/utils/retry/retry.go

```diff
@@ -0,0 +1,56 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package retry
+
+import (
+	"time"
+)
+
+// Params are used to define how retry attempts are handled.
+type Params struct {
+	MaxRetries     uint
+	InitialBackoff time.Duration
+	BackoffFactor  float64
+}
+
+// Invocation represents a function that is desired to be retried until some condition is met. It returns:
+// bool: Indicates whether the retry condition is met. False = more retries desired. True = no more retries desired.
+// interface{}: This is the returned data that the caller wants. The caller is expected to know the underlying type.
+// error: Used to indicate whether an error occurred.
+type Invocation func() (bool, interface{}, error)
```

Comment:
I do not think that we need to return `bool` and `interface{}`. It just makes things complicated.
I think that invocation function should look like below:
```
type Invocation func() error
```

Reply:
I think I could remove the `bool`, but if I remove the `interface{}` then I won't be able to return any data.

Reply:
The function needs to return at least the data and an error. In the implementation I showed you earlier (https://github.com/trustbloc/fabric-mod/blob/master/common/util/retry/retry.go) your bool is implemented in an option - WithBeforeRetry. So, if you add this function as an option then you have control of whether or not a retry is warranted. If you don't add this option then the retries keep going until success or max-retries is reached.

Your implementation is a variation on this, i.e. the bool is explicitly returned from the invocation. Not sure if we need to change it.

Reply:
You do not need to return the data. The data should be handled by the end-user. You just need en error to check whether to proceed with the retry logic or not.
The usage on the user side can be like the following:
```
func whichUsesRetry() (string,string,error) {
	var data1, data2, err

	Retry(func(){
		data1,data2,err = someFuncWith3args()
		return err
	}, &params)
	
	// some additional logic here

	return data1,data2,err
}

```

Reply:
It does not make sense to return the value. It solves nothing. How are you going to handle multiple return values?

Reply:
Ok, that will work too.

Reply:
Done!